File: benchmark_client/scripts/go/Sum of Matrices.go
package main

import (
	"encoding/binary"
	"fmt"
	"os"
	"time"
)

func readMatrix(path string) (int, []float32, error) {
	file, err := os.Open(path)
	if err != nil {
		return 0, nil, err
	}
	defer file.Close()

	var n int32
	if err := binary.Read(file, binary.LittleEndian, &n); err != nil {
		return 0, nil, err
	}
	if n <= 0 {
		return 0, nil, fmt.Errorf("invalid matrix size")
	}

	total := int(n) * int(n)
	data := make([]float32, total)
	if err := binary.Read(file, binary.LittleEndian, data); err != nil {
		return 0, nil, err
	}

	return int(n), data, nil
}

func writeMatrix(path string, n int, data []float32) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	if len(data) != n*n {
		return fmt.Errorf("invalid output matrix size")
	}

	n32 := int32(n)
	if err := binary.Write(file, binary.LittleEndian, n32); err != nil {
		return err
	}
	if err := binary.Write(file, binary.LittleEndian, data); err != nil {
		return err
	}
	return nil
}

func main() {
	n1, m1, err := readMatrix("M1.dat")
	if err != nil {
		os.Exit(1)
	}
	n2, m2, err := readMatrix("M2.dat")
	if err != nil || n1 != n2 {
		os.Exit(1)
	}

	m3 := make([]float32, n1*n1)
	start := time.Now()
	for i := range m3 {
		m3[i] = m1[i] + m2[i]
	}
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6

	if err := writeMatrix("M3.dat", n1, m3); err != nil {
		os.Exit(1)
	}
	fmt.Printf("{\"execution_time_ms\": %v}\n", elapsedMs)
}

File: benchmark_client/scripts/cpp/Sum of Matrices.cpp
#include <chrono>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

static void read_matrix(const char* filename, int& n, std::vector<float>& data) {
    std::ifstream file(filename, std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Error: File not found\n";
        exit(1);
    }

    file.read(reinterpret_cast<char*>(&n), sizeof(int));
    if (!file || n <= 0) {
        throw std::runtime_error(std::string("Failed to read matrix size from ") + filename);
    }

    const std::size_t total = static_cast<std::size_t>(n) * static_cast<std::size_t>(n);
    data.resize(total);
    file.read(reinterpret_cast<char*>(data.data()), static_cast<std::streamsize>(total * sizeof(float)));
    if (!file) {
        throw std::runtime_error(std::string("Failed to read matrix data from ") + filename);
    }
}

static void write_matrix(const char* filename, int n, const std::vector<float>& data) {
    std::ofstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error(std::string("Failed to open ") + filename);
    }

    const std::size_t total = static_cast<std::size_t>(n) * static_cast<std::size_t>(n);
    if (data.size() != total) {
        throw std::runtime_error("Invalid output matrix size.");
    }

    file.write(reinterpret_cast<const char*>(&n), sizeof(int));
    file.write(reinterpret_cast<const char*>(data.data()), static_cast<std::streamsize>(total * sizeof(float)));
}

int main() {
    try {
        int n1 = 0;
        int n2 = 0;
        std::vector<float> m1;
        std::vector<float> m2;

        read_matrix("M1.dat", n1, m1);
        read_matrix("M2.dat", n2, m2);
        if (n1 != n2) {
            return 1;
        }

        const std::size_t total = static_cast<std::size_t>(n1) * static_cast<std::size_t>(n1);
        std::vector<float> m3(total);

        const auto start = std::chrono::high_resolution_clock::now();
        for (std::size_t i = 0; i < total; ++i) {
            m3[i] = m1[i] + m2[i];
        }
        const auto end = std::chrono::high_resolution_clock::now();

        const double elapsed_ms = std::chrono::duration<double, std::milli>(end - start).count();

        write_matrix("M3.dat", n1, m3);
        std::cout << "{\"execution_time_ms\": " << elapsed_ms << "}\n";
        return 0;
    } catch (const std::exception&) {
        return 1;
    }
}
